testservices/neutronmodel/neutronmodel.go

```diff
@@ -0,0 +1,616 @@
+// neutron.SecurityGroupV2.
+func (n *NeutronModel) convertNeutronToNovaSecurityGroup(group neutron.SecurityGroupV2) nova.SecurityGroup {
+	novaGroup := nova.SecurityGroup{
+		TenantId:    group.TenantId,
+		Id:          group.Id,
+		Name:        group.Name,
+		Description: group.Description,
+		Rules:       []nova.SecurityGroupRule{},
+	}
+	var novaRules []nova.SecurityGroupRule
+	for _, rule := range group.Rules {
+		novaRules = append(novaRules, nova.SecurityGroupRule{
+			FromPort:      rule.PortRangeMin,
+			ToPort:        rule.PortRangeMax,
+			Id:            rule.Id,
+			ParentGroupId: rule.ParentGroupId,
+			IPProtocol:    rule.IPProtocol,
+			IPRange:       map[string]string{"cidr": rule.RemoteIPPrefix},
+		})
+	}
+	if len(novaRules) > 0 {
+		novaGroup.Rules = novaRules
+	}
+	return novaGroup
+}
+
+// convertNeutronToNovaSecurityGroup converts a neutron.SecurityGroupV2 to a
+// nova.SecurityGroup.
+func (n *NeutronModel) convertNovaToNeutronSecurityGroup(group nova.SecurityGroup) neutron.SecurityGroupV2 {
+	neutronGroup := neutron.SecurityGroupV2{
+		TenantId:    group.TenantId,
+		Id:          group.Id,
+		Name:        group.Name,
+		Description: group.Description,
+		Rules:       []neutron.SecurityGroupRuleV2{},
+	}
+	var neutronRules []neutron.SecurityGroupRuleV2
+	for _, rule := range group.Rules {
+		neutronRules = append(neutronRules, neutron.SecurityGroupRuleV2{
+			PortRangeMin:   rule.FromPort,
+			PortRangeMax:   rule.ToPort,
+			Id:             rule.Id,
+			ParentGroupId:  rule.ParentGroupId,
+			IPProtocol:     rule.IPProtocol,
+			RemoteIPPrefix: rule.IPRange["cidr"],
+		})
+	}
+	if len(neutronRules) > 0 {
+		neutronGroup.Rules = neutronRules
+	}
+	return neutronGroup
+}
+
+// UpdateSecurityGroup updates an existing security group given a
+// neutron.SecurityGroupRuleV2.
+func (n *NeutronModel) UpdateSecurityGroup(group neutron.SecurityGroupV2) error {
+	n.rwMu.Lock()
+	defer n.rwMu.Unlock()
+	existingGroup, err := n.SecurityGroup(group.Id)
+	if err != nil {
+		return testservices.NewSecurityGroupByIDNotFoundError(group.Id)
+	}
+	existingGroup.Name = group.Name
+	existingGroup.Description = group.Description
+	n.groups[group.Id] = *existingGroup
+	return nil
+}
+
+// UpdateNovaSecurityGroup updates an existing security group given a nova.SecurityGroup.
+func (n *NeutronModel) UpdateNovaSecurityGroup(group nova.SecurityGroup) error {
+	return n.UpdateSecurityGroup(n.convertNovaToNeutronSecurityGroup(group))
+}
+
+// AddSecurityGroup creates a new security group given a neutron.SecurityGroupV2.
+func (n *NeutronModel) AddSecurityGroup(group neutron.SecurityGroupV2) error {
+	n.rwMu.Lock()
+	defer n.rwMu.Unlock()
+	if _, err := n.SecurityGroup(group.Id); err == nil {
+		return testservices.NewSecurityGroupAlreadyExistsError(group.Id)
+	}
+	// Neutron adds 2 default egress security group rules to new security
+	// groups, copy that behavior here.
+	id, _ := strconv.Atoi(group.Id)
+	id1 := id * 999
```

Comment:
Multiply rather than mask/add for unique ids is interesting, test output might get some very large ints. :)